Partition: add key-partition document (#12963)
pingcap/docs: partitioned-table.md

## Partitioning types

This section introduces the types of partitioning in TiDB. Currently, TiDB supports [Range partitioning](#range-partitioning), [Range COLUMNS partitioning](#range-columns-partitioning), [List partitioning](#list-partitioning), [List COLUMNS partitioning](#list-columns-partitioning), [Hash partitioning](#hash-partitioning), and [Key partitioning](#key-partitioning).

> - Range partitioning, Range COLUMNS partitioning, List partitioning, and List COLUMNS partitioning are used to resolve the performance issues caused by a large number of deletions in the application, and support fast drop partition operations.

Author: ```suggestion
- Range partitioning, Range COLUMNS partitioning, List partitioning, and List COLUMNS partitioning are used to resolve the performance issues caused by a large number of deletions in the application, and support dropping partitions quickly.
```

=  1
```

### Key partitioning

> Starting from v7.0.0, TiDB supports Key partitioning. For TiDB versions earlier than v7.0.0, if you try creating a Key partitioned table, TiDB will create it as a non-partitioned table and return a warning.

Author: ```suggestion
Starting from v7.0.0, TiDB supports Key partitioning. For TiDB versions earlier than v7.0.0, if you try creating a Key partitioned table, TiDB creates it as a non-partitioned table and returns a warning.
```

Partitioning by Key requires you to append a `PARTITION BY Key (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num`, where `num` is a positive integer indicating how many partitions a table is divided into.

The following operation creates a Key partitioned table, which is divided into 4 partitions by `store_id`:

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE DEFAULT '9999-12-31',
    job_code INT,
    store_id INT

> )

Author: ```suggestion
)

```

PARTITION BY KEY(store_id)
PARTITIONS 4;
```

If `PARTITIONS num` is not specified, the default number of partitions is 1.

You can also create a Key partitioned table based on non-integer columns such as VARCHAR. For example, you can partition a table by the `fname` column:

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE DEFAULT '9999-12-31',
    job_code INT,
    store_id INT

> )

Author: ```suggestion
)

```

=  1
```

### Key partitioning

Starting from v7.0.0, TiDB supports Key partitioning. For TiDB versions earlier than v7.0.0, if you try creating a Key partitioned table, TiDB creates it as a non-partitioned table and returns a warning.

Both Key partitioning and Hash partitioning can evenly distribute data into a certain number of partitions. The difference is that Hash partitioning only supports distributing data based on a specified integer expression or an integer column, while Key partitioning supports distributing data based on a column list, and partitioning columns of Key partitioning are not limited to the integer type. The Hash algorithm of TiDB for Key partitioning is different from that of MySQL, so the table data distribution is also different.

> Partitioning by Key requires you to append a `PARTITION BY Key (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num`, where `num` is a positive integer indicating how many partitions a table is divided into.

Reviewer: ```suggestion
Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you also need to append `PARTITIONS num`, where `num` is a positive integer indicating how many partitions a table is divided into, or named definitions of the partitions like `(PARTITION pOne, PARTITION pTwo).
```

Author: @TonsnakeLin PTAL

> Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you also need to append `PARTITIONS num`, where `num` is a positive integer indicating how many partitions a table is divided into, or named definitions of the partitions like `(PARTITION pOne, PARTITION pTwo).

Author: ```suggestion
Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num` (where `num` is a positive integer indicating how many partitions a table is divided into), or add the definition of the partition names, for example, adding `(PARTITION p0, PARTITION p1)` means dividing the table into two partitions, with partition names `p0` and `p1`.
```

> Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num` (where `num` is a positive integer indicating how many partitions a table is divided into), or add the definition of the partition names (for example, adding `(PARTITION p0, PARTITION p1)` means dividing the table into two partitions, with partition names `p0` and `p1`).

Author: ```suggestion
Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num` (where `num` is a positive integer indicating how many partitions a table is divided into), or append the definition of the partition names (for example, adding `(PARTITION p0, PARTITION p1)` means dividing the table into two partitions, with partition names `p0` and `p1`).
```

> Partitioning by Key requires you to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num` (where `num` is a positive integer indicating how many partitions a table is divided into), or append the definition of the partition names (for example, adding `(PARTITION p0, PARTITION p1)` means dividing the table into two partitions, with partition names `p0` and `p1`).

Reviewer 2: ```suggestion
When partitioning by Key, you need to append a `PARTITION BY KEY (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support `GEOMETRY`). In addition, you might also need to append `PARTITIONS num` (where `num` is a positive integer indicating how many partitions a table is divided into), or append the definition of the partition names (for example, adding `(PARTITION p0, PARTITION p1)` means dividing the table into two partitions named `p0` and `p1`.
```

Both Key partitioning and Hash partitioning can evenly distribute data into a certain number of partitions. The difference is that Hash partitioning only supports distributing data based on a specified integer expression or an integer column, while Key partitioning supports distributing data based on a column list, and partitioning columns of Key partitioning are not limited to the integer type. The Hash algorithm of TiDB for Key partitioning is different from that of MySQL, so the table data distribution is also different.

Partitioning by Key requires you to append a `PARTITION BY Key (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num`, where `num` is a positive integer indicating how many partitions a table is divided into.

The following operation creates a Key partitioned table, which is divided into 4 partitions by `store_id`:

> {{< copyable "sql" >}}
> 

Author: ```suggestion
```

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE DEFAULT '9999-12-31',
    job_code INT,
    store_id INT
)
PARTITION BY KEY(store_id)
PARTITIONS 4;
```

If `PARTITIONS num` is not specified, the default number of partitions is 1.

You can also create a Key partitioned table based on non-integer columns such as VARCHAR. For example, you can partition a table by the `fname` column:

> {{< copyable "sql" >}}
> 

Author: ```suggestion
```

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE DEFAULT '9999-12-31',
    job_code INT,
    store_id INT
)
PARTITION BY KEY(fname)
PARTITIONS 4;
```

You can also create a Key partitioned table based on multiple columns. For example, you can divide a table into 4 partitions based on `fname` and `store_id`:

> {{< copyable "sql" >}}
> 

Author: ```suggestion
```

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE DEFAULT '9999-12-31',
    job_code INT,
    store_id INT
)
PARTITION BY KEY(fname, store_id)
PARTITIONS 4;
```

Currently, TiDB does not support creating Key partitioned tables if the partition column list specified in `PARTITION BY KEY` is empty. For example, after you execute the following statement, TiDB will create a non-partitioned table and return an `Unsupported partition type KEY, treat as normal table` warning.

> {{< copyable "sql" >}}
> 

Author: ```suggestion
```

- For all other statements of MySQL Linear Hash partitions, they work in TiDB the same as that in MySQL, except that the rows are distributed differently if the number of partitions is not a power of 2, which will give different results for [partition selection](#partition-selection), `TRUNCATE PARTITION`, and `EXCHANGE PARTITION`.

### How TiDB handles Linear Key partitions

Starting from v7.0.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR HASH` syntax for Key partitioning. However, TiDB ignores the `LINEAR` keyword and uses a non-linear hash algorithm instead.

> Before v7.0.0, if you try creating a Key partitioned table, TiDB will create it as a non-partitioned table and return a warning.

Author: ```suggestion
Before v7.0.0, if you try creating a Key partitioned table, TiDB creates it as a non-partitioned table and returns a warning.
```

- For all other statements of MySQL Linear Hash partitions, they work in TiDB the same as that in MySQL, except that the rows are distributed differently if the number of partitions is not a power of 2, which will give different results for [partition selection](#partition-selection), `TRUNCATE PARTITION`, and `EXCHANGE PARTITION`.

### How TiDB handles Linear Key partitions

> Starting from v7.0.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR HASH` syntax for Key partitioning. However, TiDB ignores the `LINEAR` keyword and uses a non-linear hash algorithm instead.

Author: ```suggestion
Starting from v7.0.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR KEY` syntax for Key partitioning. However, TiDB ignores the `LINEAR` keyword and uses a non-linear hash algorithm instead.
```

ERROR 8200 (HY000): Unsupported optimize partition
```

### Key partition management

Currently, Key partitioning only supports the `ALTER TABLE ... TRUNCATE PARTITION` partition management statement.

> 

Author: ````suggestion
```sql
ALTER TABLE members TRUNCATE PARTITION p0;
```

```
Query OK, 0 rows affected (0.03 sec)
````
```

> If you execute a Key partition management statement that is not yet supported, TiDB will return an error.

Reviewer 3: Please limit the use of "will" and try to use the present tense.
```suggestion
If you execute a Key partition management statement that is not yet supported, TiDB returns an error.
```

Author: Thanks for the comments. The other places are updated in https://github.com/pingcap/docs/pull/12963/commits/00def2a14f5a99297b89e783528c880f01e40c21.

### Compatibility with MySQL

> Currently, TiDB supports Range partitioning, Range COLUMNS partitioning, List partitioning, List COLUMNS partitioning, Hash partitioning, and Key partitioning. Other partitioning types that are available in MySQL are not supported yet in TiDB.

Author: Good catch. Added a comment in https://github.com/pingcap/docs-cn/pull/13223/files#r1141761876

> Currently, TiDB supports Range partitioning, Range COLUMNS partitioning, List partitioning, List COLUMNS partitioning, Hash partitioning, and Key partitioning. Other partitioning types that are available in MySQL are not supported yet in TiDB.

Author: 中文已更新 https://github.com/pingcap/docs-cn/pull/13223/commits/ee043bc87f85da621ab974669b63005f0751ec1d
